File: WithCallback/app/src/main/java/com/ducluanxutrieu/tikiapp/ui/home/HomeViewModelFactory.java
package com.ducluanxutrieu.tikiapp.ui.home;

import androidx.annotation.NonNull;
import androidx.lifecycle.ViewModel;
import androidx.lifecycle.ViewModelProvider;

import com.ducluanxutrieu.tikiapp.data.RetrofitClient;
import com.ducluanxutrieu.tikiapp.data.RetrofitService;
import com.ducluanxutrieu.tikiapp.data.TikiDatabase;
import com.ducluanxutrieu.tikiapp.data.home.BannerDao;
import com.ducluanxutrieu.tikiapp.data.home.HomeRepository;

public class HomeViewModelFactory implements ViewModelProvider.Factory {
    public HomeViewModelFactory(TikiDatabase database) {
        this.database = database;
    }

    private final TikiDatabase database;

    @NonNull
    @Override
    public <T extends ViewModel> T create(@NonNull Class<T> modelClass) {
        BannerDao bannerDao = database.bannerDao();
        RetrofitService networkService = RetrofitClient.getRetrofitInstance();

        if (modelClass.isAssignableFrom(HomeViewModel.class)){
            return (T) new HomeViewModel(new HomeRepository(bannerDao, networkService));
        }

        throw new IllegalArgumentException("Unknown ViewModel class");
    }
}

File: WithCoroutines/app/src/main/java/com/example/tikiapp/ui/home/HomeViewModelFactory.kt
package com.example.tikiapp.ui.home

import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider
import com.example.tikiapp.data.home.HomeRepository
import com.example.tikiapp.data.home.TikiDatabase
import com.example.tikiapp.util.getNetworkService

/**
 * ViewModel provider factory to instantiate LoginViewModel.
 * Required given LoginViewModel has a non-empty constructor
 */
class HomeViewModelFactory(private val database: TikiDatabase) : ViewModelProvider.Factory {

    @Suppress("UNCHECKED_CAST")
    override fun <T : ViewModel> create(modelClass: Class<T>): T {
        if (modelClass.isAssignableFrom(HomeViewModel::class.java)) {
            return HomeViewModel(
                    HomeRepository(getNetworkService(), database.bannerDao)
            ) as T
        }
        throw IllegalArgumentException("Unknown ViewModel class")
    }
}